### PR TITLE
[NNAPI EP] Add Gather op support

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
**Description**:

- Added Gather op support for NNAPI EP
- Modified several UT for gather op test to enable testing for nnapi ep

**Motivation and Context**

MobileBERT missing operator support

### CODE REVIEW

**Review by skottmckay — DISMISSED** [review]
:shipit:

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc [hunk 1]
```diff
@@ -2436,6 +2436,70 @@ Status FlattenOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, cons
...
+  NodeAttrHelper helper(node_unit);
+  int32_t rank = static_cast<int32_t>(shaper[input1].size());
+  int32_t axis = static_cast<int32_t>(HandleNegativeAxis(helper.Get("axis", 0), rank));
+
+  std::vector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(input1));
+  ADD_SCALAR_OPERAND(model_builder, input_indices, axis);
+
+  // Add indices operand into nnapi
+  const auto& indices_tensor = *initializers.at(node_unit.Inputs()[1].node_arg.Name());
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2473]:
```suggestion
  const auto& indices_tensor = *initializers.at(input2);
```

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc [hunk 2]
```diff
@@ -2436,6 +2436,70 @@ Status FlattenOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, cons
...
+  uint32_t size = 1;
+  Shape indices_dimen;
+  indices_dimen.reserve(indices_tensor.dims_size());
+  for (auto i = 0; i < indices_tensor.dims_size(); i++) {
+    size *= indices_shape[i];
+    indices_dimen.push_back(static_cast<uint32_t>(indices_shape[i]));
+  }
+  std::vector<int32_t> indices(size);
+  for (uint32_t i = 0; i < size; i++) {
+    indices[i] = SafeInt<int32_t>(raw_indices_data[i]);
```
Comment (edgchen1) [inline onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2488]: I'm not so sure about using reinterpret_cast for raw_indices_data. It's kind of tricky to use without undefined behavior.
An alternative:
```suggestion
    int64_t index_i64;
    memcpy(&index_i64, unpacked_tensor.data() + i * sizeof(int64_t), sizeof(int64_t));
    indices[i] = SafeInt<int32_t>(index_i64);
```

https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8#type-punning-arrays

Comment (YUNQIUGUO) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc:2488]: Thanks for asking this. It is necessary to consider both int32/int64 type indices here. The unit tests do not cover a case with float input data + int32 indices,  overlooked it before.

Do you have a suggested good way to accommodate both cases here?

Comment (YUNQIUGUO) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc]: Added additional unit test to separately to cover int32_t indices case with float input data for NNAPI.

File: onnxruntime/test/providers/cpu/tensor/gather_op_test.cc [hunk 3]
```diff
@@ -107,78 +125,154 @@ TEST(GatherOpTest, Gather_invalid_index_gpu) {
...
+                          10.0f, 10.1f, 10.2f, 10.3f,
+                          11.0f, 11.1f, 11.2f, 11.3f,
+                          12.0f, 12.1f, 12.2f, 12.3f});
+    test.AddInput<int32_t>("indices", {1}, {1}, indices_is_initializer);
+    test.AddOutput<float>("output", {1, 3, 4},
+                          {10.0f, 10.1f, 10.2f, 10.3f,
+                           11.0f, 11.1f, 11.2f, 11.3f,
+                           12.0f, 12.1f, 12.2f, 12.3f});
+    // NNAPI EP only supports float input data for now,
+    // this test case separately tests int32_t indices case with Nnapi
```
Comment (edgchen1) [inline onnxruntime/test/providers/cpu/tensor/gather_op_test.cc:245]: thanks for adding more tests. do the int32 indices tests need to be limited to NNAPI EP?

Comment (YUNQIUGUO) [reply onnxruntime/test/providers/cpu/tensor/gather_op_test.cc]: um, for sure these tests can also run on cpu. removed the condition.

File: onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc (hunk 2 above)
Comment (skottmckay) [reply onnxruntime/core/providers/nnapi/nnapi_builtin/builders/op_builder.cc]: Can we reference the gist in a comment here to explain the use of memcpy?

**Review by skottmckay — APPROVED** [review]
:shipit:

**Review by edgchen1 — APPROVED** [review]

**Pull request merged**